src/cloud-providers/gcp/provider.go

```diff
@@ -0,0 +1,191 @@
+// (C) Copyright Confidential Containers Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+package gcp
+
+import (
+	"context"
+	"encoding/base64"
+	"fmt"
+	"log"
+	"net/netip"
+
+	compute "cloud.google.com/go/compute/apiv1"
+	computepb "cloud.google.com/go/compute/apiv1/computepb"
+	provider "github.com/confidential-containers/cloud-api-adaptor/src/cloud-providers"
+	"github.com/confidential-containers/cloud-api-adaptor/src/cloud-providers/util"
+	"github.com/confidential-containers/cloud-api-adaptor/src/cloud-providers/util/cloudinit"
+	"golang.org/x/oauth2/google"
+	option "google.golang.org/api/option"
+	proto "google.golang.org/protobuf/proto"
+)
+
+var logger = log.New(log.Writer(), "[adaptor/cloud/gcp] ", log.LstdFlags|log.Lmsgprefix)
+var computeScope = "https://www.googleapis.com/auth/compute"
+
+const maxInstanceNameLen = 63
+
+type gcpProvider struct {
+	serviceConfig   *Config
+	instancesClient *compute.InstancesClient
+}
+
+func (p *gcpProvider) ConfigVerifier() error {
+	return nil
+}
+
+func NewProvider(config *Config) (provider.Provider, error) {
+	logger.Printf("gcp config: %#v", config.Redact())
+	provider := &gcpProvider{
+		serviceConfig:   config,
+		instancesClient: nil,
+	}
+	if config.GcpCredentials != "" {
+		creds, err := google.CredentialsFromJSON(context.TODO(), []byte(config.GcpCredentials), computeScope)
+		if err != nil {
+			return nil, fmt.Errorf("configuration error when using creds: %s", err)
+		}
+		provider.instancesClient, err = compute.NewInstancesRESTClient(context.TODO(), option.WithCredentials(creds))
+		if err != nil {
+			return nil, fmt.Errorf("NewInstancesRESTClient with credentials error: %s", err)
+		}
+	} else {
```

Comment:
Is having an empty GcpCredentials a valid case?

Reply:
Yes, depending on the deployment environment, the library can use ADC. I haven't deeply explored it, though.